src/index.js

```diff
@@ -40,40 +41,72 @@ function glamorous(comp, {rootEl, displayName} = {}) {
      *   merged predicably (with no regard to specificity)
      * @return {ReactElement} React.createElement
      */
-    function GlamorousComponent(props) {
-      const {className, ...rest} = props
-      const {toForward, cssOverrides} = splitProps(rest, GlamorousComponent)
-
-      // create className to apply
-      const mappedArgs = GlamorousComponent.styles.map(mapToPropsCall)
-      const {
-        glamorStyles: parentGlamorStyles,
-        glamorlessClassName,
-      } = extractGlamorStyles(className)
-      const glamorClassName = css(
-        ...mappedArgs,
-        ...parentGlamorStyles,
-        cssOverrides,
-      ).toString()
-      const fullClassName = joinClasses(glamorlessClassName, glamorClassName)
-
-      return React.createElement(GlamorousComponent.comp, {
-        className: fullClassName,
-        ...toForward,
-      })
-
-      function mapToPropsCall(glamorRules) {
-        if (typeof glamorRules === 'function') {
-          return glamorRules(props)
+    class GlamorousComponent extends Component {
+      constructor(props, context) {
+        super(props, context)
+        this.state = {
+          theme: null,
         }
-        return glamorRules
+        this.setTheme = this.setTheme.bind(this)
+      }
+
+      setTheme(theme) {
+        this.setState({
+          theme,
+        })
+      }
+
+      componentWillMount() {
+        if (this.context[CHANNEL]) {
+          // subscribe to future theme changes
+          this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme)
+        } else {
+          this.setTheme(this.props.theme || {})
+        }
+      }
+
+      componentWillUnmount() {
+        // cleanup subscription
+        this.unsubscribe && this.unsubscribe()
+      }
+
+      render() {
+        const {className, ...rest} = this.props
+        const {toForward, cssOverrides} = splitProps(rest, GlamorousComponent)
+        // create className to apply
+        const mappedArgs = GlamorousComponent.styles.map(glamorRules => {
+          if (typeof glamorRules === 'function') {
+            return glamorRules(this.props, {...this.state.theme})
+          }
+          return glamorRules
+        })
+        const {
+          glamorStyles: parentGlamorStyles,
+          glamorlessClassName,
+        } = extractGlamorStyles(className)
+        const glamorClassName = css(
+          ...mappedArgs,
+          ...parentGlamorStyles,
+          cssOverrides,
+        ).toString()
+        const fullClassName = joinClasses(glamorlessClassName, glamorClassName)
+
+        return React.createElement(GlamorousComponent.comp, {
+          className: fullClassName,
+          ...toForward,
+        })
       }
     }
+
     GlamorousComponent.propTypes = {
       className: PropTypes.string,
       cssOverrides: PropTypes.object,
+      theme: PropTypes.object,
```

Comment:
This is exactly what I had in mind.
In case a prop `theme` is provided, it takes precedence over context.
I would not even merge with context in this case.